[expo-updates][iOS] Logging and log reading

douglowder opened the pull request:
# Why

Provides a way to log from both Objective-C and Swift code in expo-updates, and provides code to read all logs written since a given time.

# How

- New Swift code implementing the logger, log reader, expo-updates specific error codes, and JSON format for log messages
- Logs with severity "error" or "fatal" will include stack traces (limited to 20 frames)
- Leverages the existing Logger in expo-modules-core.

# Test Plan

New unit tests provided for both Swift and Objective-C.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

douglowder committed:
[expo-updates] Logging and log reading

Why

Provides a way to log from both Objective-C and Swift code in expo-updates, and provides code to read all logs written since a given time.

How

- New Swift code implementing the logger, log reader, expo-updates specific error codes, and JSON format for log messages
- Logs with severity "error" or "fatal" will include stack traces

Test Plan

New unit tests provided for both Swift and Objective-C.

esamelson commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,59 @@
+
+      // Filter the log to select our subsystem and the expo-updates category
+      // and extract just the log message strings
+      let allEntriesMessages = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
+          .filter { $0.subsystem == "dev.expo.modules" }
+          .filter { $0.category == "expo-updates" }
+          .compactMap { $0.composedMessage }
```
It looks like we're reading all entries into memory and _then_ filtering them; do you think performance is something worth being concerned about here? (Should we put an upper bound on the `newerThan` date, for example?)

esamelson replied:
I was thinking more along the lines of enforcing a limit in this method, just to keep us from shooting ourselves in the foot in the future? But if you would rather keep the flexibility (or just don't think it will be an issue) that's totally fine w me too

douglowder replied:
I did two things here:
- only allow logs to be searched going back 1 day
- added a predicate to the OSLogStore call so we only retrieve the entries we want and don't have to filter after the fact

esamelson commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,43 @@
+
+import Foundation
+
+public struct UpdatesLogEntry: Codable {
```
is it possible to expose this to obj-c if it also subclasses NSObject? Seems like it would be nice if the class calling `getLogEntries` weren't also responsible for parsing the entries, but also understandable if it's too tricky to implement across both languages.

douglowder replied:
Sure I can have a look

douglowder replied:
Exposing this to Objective C would be a complete reimplementation and be a bit awkward. It might be better to change the reader to return a list of `NSDictionary` objects for Objective-C usage.

esamelson replied:
Cool -- that makes sense to me. Thanks for checking that out.

douglowder replied:
Now Objective-C can call the reader and get entries returned as NSDictionaries.

esamelson commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:
```diff
@@ -0,0 +1,42 @@
+  case NoUpdatesAvailable = 1
+  case UpdateAssetsNotAvailable = 2
+  case UpdateServerUnreachable = 3
+  case UpdateHasIncorrectHash = 4
```
it would probably be more terminologically correct to say `UpdateHasInvalidSignature` as we don't have "hashes" at the update level

douglowder replied:
Fixed

esamelson reviewed (commented):
Cool, this looks like it will be nice to use! I think there will probably be more error codes to add over time -- one that comes to mind is `AssetHasMismatchedHash` or something to that effect -- and it's possible we might not end up using all the ones there now, but probably fine to change/grow the list as needed.

esamelson reviewed (approved):
LGTM, excited to see this in action!

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift:
```diff
@@ -0,0 +1,159 @@
+@objc(EXUpdatesLogger)
+public class UpdatesLogger: NSObject {
+
+  private let osLogger: os.Logger
+
+  public override init() {
+    osLogger = os.Logger(subsystem: "dev.expo.modules", category: "expo-updates")
+  }
```
```suggestion
  private let osLogger = os.Logger(subsystem: "dev.expo.modules", category: "expo-updates")
```

douglowder replied:
Done

wschurman commented on packages/expo-updates/ios/Tests/EXUpdatesLoggerOCTests.m:
```diff
@@ -0,0 +1,84 @@
+@end
+
+// Headers for the Swift classes and enum
+// In application code, EXUpdates-Swift.h can be imported instead
```
I wonder why this isn't visible to tests. It's been a few months since I did swift stuff so I don't remember if I ever got this bridge working.

That being said, I think just having the swift test is sufficient. We can rely upon the compiler to tell us if the bridge is broken.

douglowder replied:
Replaced the OC test with a Swift test to verify the dictionary reader method

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,76 @@
+        result.add(logEntry!.asDict())
+      }
+    } catch {
+      result.add("Error occurred in UpdatesLogReader: \(error.localizedDescription)")
```
not sure this pattern of returning an error string as an arbitrary element will be consumable. Might make more sense to just throw the error?

This also is a bit messy for the dictionary element return type.

douglowder replied:
Hmm I will think more on how to handle this better. No error should ever be thrown really, but I'll have a look at this.

douglowder replied:
Completed this change

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,76 @@
+   Maximum of one day lookback is allowed
+   */
+  @objc(getLogEntriesNewerThan:returnEntriesAs:)
+  public func getLogEntries(newerThan: Date, returnEntriesAs: String = "NSDictionary") -> NSArray {
```
Prefer two methods that have more concrete return types than a dynamic return type

douglowder replied:
Done

douglowder committed:
Use ExpoModulesCore.Logger for expo-updates logs

- Make Logger subclassable with one overridable method
- Add defined constants for log subsystem and categories

douglowder committed:
Address review comments

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:
```diff
@@ -0,0 +1,42 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+// Error codes for expo-updates logs
```
Done

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,76 @@
+public class UpdatesLogReader: NSObject {
+
+  @objc(getLogEntriesNewerThan:)
+  public func getLogEntries(newerThan: Date) -> NSArray {
```
Done

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,61 @@
+    }
+  }
+
+  public func asDict() -> NSDictionary {
```
Done

tsapeta reviewed (changes requested):
I quickly went through the PR so it's just the first iteration

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,83 @@
+ */
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+@objc(EXUpdatesLogReader)
+public class UpdatesLogReader: NSObject {
```
As written, this class is quite specific to `expo-updates`. However, I agree that in future we could split the more general parts of this class out to something more generic that lives in `expo-modules-core`.

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,83 @@
+    let strings: [String] = getLogEntries(newerThan: newerThan)
+    var result: [[String: Any]] = []
+    strings.forEach { s in
+      let entry = UpdatesLogEntry.create(from: s)
+      if (entry != nil) {
+        let dict = entry?.asDict()
+        if (dict != nil) {
+          result.append(dict!)
+        }
+      }
```
Good point about `compactMap`, I'll take a look at that method

douglowder replied:
Fixed

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,83 @@
+  /**
+   Convenience method for returning expo-updates logs from last hour
+   */
+  @objc public func logEntriesInLastHour() -> [String] {
```
No they were intended to be public. I'll think about whether we still need these or not, though.

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift:
```diff
@@ -0,0 +1,177 @@
+/**
+ Subclass of Logger that overrides the format of log messages
+ */
+internal class CoreLogger: ExpoModulesCore.Logger {
```
The intent of this feature was to log messages in JSON format. What I can do instead is remove this change, and handle the prefix in `UpdatesLogReader` during message extraction.

douglowder replied:
Fixed

douglowder commented on packages/expo-modules-core/ios/Swift/Logging/Logger.swift:
```diff
@@ -2,9 +2,13 @@
+
+open class Logger {
+
+  public static let LOG_SUBSYSTEM = "dev.expo.modules"
+  public static let LOG_CATEGORY = "expo"
```
Good point, I'll make this change

douglowder replied:
Fixed

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,83 @@
+    var result: [String] = []
+    do {
+      let earliestDate = Date().addingTimeInterval(-86400)
+      let dateToUse = newerThan.timeIntervalSince1970 < earliestDate.timeIntervalSince1970 ?
+      earliestDate :
+      newerThan
```
Fixed

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,83 @@
+  /**
+   Convenience method for returning expo-updates logs from last day
+   */
+  @objc public func logEntriesInLastDay() -> [String] {
```
Removed this

douglowder committed:
Address review comments

- Back out changes to subclass Logger
- Remove unneeded methods
- Fix SwiftLint warnings

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,76 @@
+   Convenience method for returning expo-updates logs from last day
+   */
+  @objc public func logEntriesInLastDay() -> NSArray {
+    return getLogEntries(newerThan: Date().addingTimeInterval(-86400))
```
I'm removing these as I don't think they are really needed

douglowder committed:
Additional changes

- Remove OC test class and replace with Swift class to test reading as dictionaries
- Limit stacktraces to 20 frames and consolidate implementation
- If OSLogReader fails, emit the error in UpdatesLogEntry format

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,63 @@
+      // Extract just the log message strings, removing the first two characters added
+      // by ExpoModulesCore.Logger
+      result = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
```
remove

douglowder replied:
Done

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,63 @@
+      // by ExpoModulesCore.Logger
+      result = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
+          .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
```
this can just be `map`

douglowder replied:
I'd prefer to leave it as `compactMap`

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,78 @@
+    do {
+      let jsonDecoder = JSONDecoder()
+      guard let jsonData = from.data(using: .utf8) else { return nil }
+      let logEntry: UpdatesLogEntry = try jsonDecoder.decode(UpdatesLogEntry.self, from: jsonData)
+      return logEntry
```
```suggestion
      return try jsonDecoder.decode(UpdatesLogEntry.self, from: jsonData)
```

wschurman reviewed (approved):
Just a few more changes before commit. May be worth discussing the error case mentioned in the inline comment.

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,63 @@
+import OSLog
+
+import ExpoModulesCore
+import SafariServices
```
Done

douglowder commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,63 @@
+          .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
+          }
+    } catch {
+      // If an error happened reading the OSLogStore, surface the error as an UpdatesLogEntry with the right shape
```
Ok modified this to throw instead, and made appropriate adjustments to test code

douglowder committed:
Address additional review comments

wschurman commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,77 @@
+    result["message"] = message
+    result["code"] = code
+    result["level"] = level
+    result["updateId"] = updateId != nil ? updateId : NSNull()
+    result["assetId"] = assetId != nil ? assetId : NSNull()
+    if stacktrace != nil {
+      let nsstack = NSMutableArray()
+      for s in stacktrace! {
+        nsstack.add(s)
+      }
+      result["stacktrace"] = nsstack
+    } else {
+      result["stacktrace"] = NSNull()
```
I believe we can use swift types here, and they'll be automatically converted over the bridge (for free)

douglowder replied:
Sadly this does not seem to be the case. Passing in `nil` in Swift does not show up as an NSNull in OC.

Perhaps a cleaner way is to simply not have the keys present in the dictionary for nulls.

douglowder replied:
Fixed

wschurman reviewed (approved):
One more small comment but otherwise looks good!

douglowder committed:
Use Swift types in LogEntry dictionaries

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,79 @@
+    result["message"] = message
+    result["code"] = code
+    result["level"] = level
+    if (updateId != nil) {
+      result["updateId"] = updateId
+    }
+    if (assetId != nil) {
+      result["assetId"] = assetId
+    }
```
Generally, we don't use `()` in ifs, switches, etc.
Is there any reason why you check if it's non-nil? I think you can just assign it and don't care 😅

douglowder replied:
I did this to ensure that OC code got the dictionaries with the right behavior. I'll check to see if that is still necessary.

tsapeta replied:
Since this is Swift dictionary, assigning `nil` means nothing else than just deleting the key from the dictionary if it existed

douglowder replied:
Fixed the syntax

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,79 @@
+      result["assetId"] = assetId
+    }
+    if stacktrace != nil {
+      let nsstack = NSMutableArray()
```
Why `NSMutableArray`? Please use Swift types.

douglowder replied:
Fixed

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+    if stacktrace != nil {
+      let nsstack = NSMutableArray()
+      for s in stacktrace! {
```
I think I already mentioned this somewhere — never use force unwrapping `!` except in tests and some edge-cases where it makes sense and we're sure it won't crash. Here the situation is simple, just do
```
if let stacktrace = stacktrace {
```
instead of
```
if stacktrace != nil {
```
to safely unwrap the variable

douglowder replied:
That is an interesting syntax

douglowder replied:
Fixed

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+  }
+
+  /**
+   Utility method to construct stacktrace as a string array for log entries
+   */
+  private static let STACKTRACE_MAX_LENGTH = 20
```
I guess you meant to put this comment a few lines below to refer to `currentStackTrace` instead of the `STACKTRACE_MAX_LENGTH` constant 😛

douglowder replied:
Fixed

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:
```diff
@@ -0,0 +1,54 @@
+
+    // Extract just the log message strings, removing the first two characters added
+    // by ExpoModulesCore.Logger
+    return allEntries
+        .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
+        }
```
```suggestion
    return allEntries
      .map { entry in
        let suffixFrom = entry.composedMessage.index(entry.composedMessage.startIndex, offsetBy: 2)
        return String(entry.composedMessage.suffix(from: suffixFrom))
      }
```

douglowder replied:
Fixed

douglowder committed:
Address review comments

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:
```diff
@@ -0,0 +1,43 @@
+  // the usual "\(UpdatesErrorCode.NoUpdatesAvailable)"
+  // string representation will not work as expected,
+  // so we add this representation here
+  public var asString: String {
```
I've got an idea, maybe we can make the enum be a String instead of Int? Then I think we won't have to map it to string 🤔

douglowder replied:
Funny story, that's exactly what I did on Android. However, I think it's better this way here as an integer enum is more usual in Objective C.

tsapeta commented on packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:
```diff
@@ -0,0 +1,79 @@
+      result["assetId"] = assetId
+    }
+    if let stacktrace = stacktrace {
+      var nsstack = [String]()
+      for s in stacktrace {
+        nsstack.append(s)
+      }
+      result["stacktrace"] = nsstack
```
Isn't it the same? 😉 
```suggestion
      result["stacktrace"] = stacktrace
```

douglowder replied:
Ah yes, it was not the same when I was doing NSArray. I'll fix this.

tsapeta commented on packages/expo-updates/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Logging and log reading. ([#18284](https://github.com/expo/expo/pull/18284) by [@douglowder](https://github.com/douglowder))
```
Could you make it more descriptive? Also, it's not a new feature from the user perspective — it doesn't change the public API and doesn't bring any visible value for the user, so I would prefer moving this to "Others" category.
This reminded me that I should clarify it in our contributing guide about updating changelogs 😅

douglowder replied:
True -- and the expotools add-changelog should support "Others" as well :)

douglowder committed:
Address review comments
